crates/turborepo-lib/src/global_deps_package_detector.rs

```diff
@@ -0,0 +1,146 @@
+use thiserror::Error;
+use turbopath::AnchoredSystemPath;
+use turborepo_repository::{
+    change_mapper::{DefaultPackageDetector, PackageDetection, PackageDetector},
+    package_graph::{PackageGraph, WorkspacePackage},
+};
+use wax::{BuildError, Program};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    InvalidFilter(#[from] BuildError),
+}
+
+/// A package detector that uses a global deps list to determine
+/// if a file should cause all packages to be marked as changed.
+/// This is less conservative than the `DefaultPackageDetector`
+/// which assumes that any changed file that is not in a package
+/// is a root dependency. Since we have a list of global deps,
+/// we can check against that and avoid invalidating in unnecessary cases.
+pub struct GlobalDepsPackageDetector<'a> {
+    pkg_dep_graph: &'a PackageGraph,
+    global_deps_matcher: wax::Any<'a>,
+}
+
+impl<'a> GlobalDepsPackageDetector<'a> {
+    pub fn new<S: wax::Pattern<'a>, I: Iterator<Item = S>>(
+        pkg_dep_graph: &'a PackageGraph,
+        global_deps: I,
+    ) -> Result<Self, Error> {
+        let global_deps_matcher = wax::any(global_deps)?;
+
+        Ok(Self {
+            pkg_dep_graph,
+            global_deps_matcher,
+        })
+    }
+}
+
+impl<'a> PackageDetector for GlobalDepsPackageDetector<'a> {
+    fn detect_package(&self, path: &AnchoredSystemPath) -> PackageDetection {
+        match DefaultPackageDetector::new(self.pkg_dep_graph).detect_package(path) {
+            // Since `DefaultPackageDetector` is overly conservative, we can check here if
+            // the path is actually in globalDeps and if not, return it as
+            // PackageDetection::Package(WorkspacePackage::root()).
+            PackageDetection::All => {
+                let cleaned_path = path.clean();
+                let in_global_deps = self.global_deps_matcher.is_match(cleaned_path.as_str());
```

Comment:
I'm having trouble tracing this code. Does the `global_deps_matcher` always get the default global deps of `package.json`/`turbo.json` added to it? Otherwise we'll incorrectly map a `package.json` change just be a root package change instead of a all change.

Or I'm misunderstanding these changes completely and this codepath isn't hit in that scenario.

Edit: I was misunderstanding, default global deps are checked at `ChangeMapper` which uses a package detector if a default global change isn't detected